components/brave_wallet/browser/keyring_controller.cc

```diff
@@ -675,6 +730,108 @@ std::vector<mojom::AccountInfoPtr> KeyringController::GetAccountInfosForKeyring(
   return result;
 }
 
+void KeyringController::GetHardwareAccounts(
+    GetHardwareAccountsCallback callback) {
+  std::vector<mojom::AccountInfoPtr> accounts;
+  base::Value hardware_keyrings(base::Value::Type::DICTIONARY);
+  const base::Value* value = GetPrefForHardwareKeyring(prefs_);
+  if (!value) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  for (const auto hw_keyring : value->DictItems()) {
+    std::string device_id = hw_keyring.first;
+    const base::Value* account_value = hw_keyring.second.FindKey(kAccountMetas);
+    if (!account_value)
+      continue;
+    SerializeHardwareAccounts(account_value, &accounts);
+  }
+
+  std::move(callback).Run(std::move(accounts));
+}
+
+// static
+const base::Value* KeyringController::GetPrefForHardwareKeyring(
+    PrefService* prefs) {
+  DCHECK(prefs);
+  const base::DictionaryValue* keyrings_pref =
+      prefs->GetDictionary(kBraveWalletKeyrings);
+  if (!keyrings_pref)
+    return nullptr;
+  return keyrings_pref->FindKey(kHardwareKeyrings);
+}
+
+void KeyringController::AddHardwareAccounts(
+    std::vector<mojom::HardwareWalletAccountPtr> infos) {
+  if (infos.empty())
+    return;
+  const auto& hardware_vendor = infos.front()->hardware_vendor;
+  const auto hash = base::PersistentHash(infos.front()->address);
+  std::string device_id = hardware_vendor + std::to_string(hash);
+
+  base::Value hardware_keyrings(base::Value::Type::DICTIONARY);
+  const base::Value* value = GetPrefForHardwareKeyring(prefs_);
+  if (value) {
+    hardware_keyrings = value->Clone();
+  }
+  const base::Value* device_value = hardware_keyrings.FindKey(device_id);
+  base::Value device_keyring(base::Value::Type::DICTIONARY);
+  if (device_value)
+    device_keyring = device_value->Clone();
+
+  const base::Value* meta_value = device_keyring.FindKey(kAccountMetas);
+  base::Value account_meta(base::Value::Type::DICTIONARY);
+  if (meta_value)
+    account_meta = meta_value->Clone();
+
+  for (const auto& info : infos) {
+    DCHECK_EQ(hardware_vendor, info->hardware_vendor);
+    if (hardware_vendor != info->hardware_vendor)
+      continue;
+    base::Value hw_account(base::Value::Type::DICTIONARY);
+    hw_account.SetStringKey(kAccountName, info->name);
+    hw_account.SetStringKey(kHardwareVendor, info->hardware_vendor);
+    hw_account.SetStringKey(kHardwareDerivationPath, info->derivation_path);
+
+    account_meta.SetKey(info->address, std::move(hw_account));
+  }
+
+  device_keyring.SetKey(kAccountMetas, std::move(account_meta));
+  hardware_keyrings.SetKey(device_id, std::move(device_keyring));
+
+  DictionaryPrefUpdate update(prefs_, kBraveWalletKeyrings);
```

Comment:
we can have ` base::Value* GetPrefForHardwareKeyring(PrefService* prefs)` which uses `DictionaryPrefUpdate` and return a mutable `kHardwareKeyrings` so we don't have to do multiple `Clone` and set the value in the end, we can modify the `base::Value*` directly instead

Reply:
like this
https://github.com/brave/brave-core/pull/10102/commits/10bd683f95363d743b2fb72a09278f0b5c3bb92e#diff-b44d65fd43bf18c3c2eba3fe2c9d553db219bb2a34b3762cfd8b990f8fe7aa68R175-R193

Reply:
done